src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -1020,6 +1020,34 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
             parseError(s"method `$method' expects a lambda function [param => Any], got invalid mapping [param => $error]")
         }
 
+      case (it: TIterable, "flatMap", rhs) =>
+        lhs.typecheck(ec)
+        val (param, body) = rhs match {
+          case Array(Lambda(_, p, b)) => (p, b)
+          case _ => parseError(s"method `$method' expects a lambda function [param => Any]")
+        }
+        body.typecheck(ec.copy(st = ec.st + ((param, (-1, it.elementType)))))
+        `type` = body.`type` match {
+          case t: TIterable => (t, it) match {
```

Comment:
I did this to give a more precise error message in the case where the body was also iterable but of a different type. Relates to comment above, regarding consistency of array / array and set / set.  If you want support for lhs array and expr set, returning a set, then I'll modify further, but I still think a more precise error message is helpful as flatMap is a bit subtle.